README.md
# SatMAD Applications and Examples

[![Documentation Status](https://readthedocs.org/projects/satmad-applications/badge/?version=latest)](https://satmad-applications.readthedocs.io/en/latest/?badge=latest)
[![DOI](https://zenodo.org/badge/351173260.svg)](https://zenodo.org/badge/latestdoi/351173260)

SatMAD is an open source Python package, aiming at providing the base functionality to solve satellite mission analysis
and design as well as orbital mechanics problems with enough precision and performance to be used in the design and
operation of real satellites.

This project contains applications, analyses and examples with SatMAD project.

## Documentation

The documentation for SatMAD is here: <https://satmad.readthedocs.io/>

The documentation of SatMAD-Applications with all the analyses and examples are here:
<https://satmad-applications.readthedocs.io>

The documentation contains text-only version of these analyses. For more hands-on Jupyter notebooks, you can try the
github address of the 
[satmad-applications project](https://github.com/egemenimre/satmad_applications/tree/main/docs/analyses).

## Installing SatMAD

The SatMAD package is on [PyPI](https://pypi.org/project/satmad/) and you can install it simply by running:

    pip install satmad

You can also install it via [conda-forge](https://github.com/conda-forge/satmad-feedstock):

    conda install -c conda-forge satmad

Do not install satmad using sudo.

You can find the source code on GitHub: <https://github.com/egemenimre/satmad>

## Requirements

-   SatMAD provides most of the orbital mechanics routines for the analyses
-   NumPy and SciPy are used for the underlying mathematical algorithms
-   Matplotlib is used for plots
-   Astropy handles all time and reference frame computations

## License

This project is Copyright (c) Egemen Imre and licensed under the terms of the GNU GPL v3+ license.


docs/changelog.md
# Changelog

Changes and additions to the examples and analyses are given here.

## Latest Version

### Version 0.2 (16 May 2021)

- Migrated to [MyST](https://myst-parser.readthedocs.io/en/latest/) document parser
- Added [Groundstation communication times analysis](how_to_guides/basics/gs_access.ipynb)
- Added [Hohmann Transfer](how_to_guides/maneuver/hohmann_1.ipynb)

### Under development: Version 0.3

- Added Binder links
- Added [Sentinel-2 Orbit Evolution](analyses/sats/s2_analysis.ipynb)


## Previous Versions


### Version 0.1 (25 Mar 2021)

- Initialised the project with three analyses (energy conservation properties of scipy diff solvers in the context of
  numerical propagation)





satmad_applications/analyses/propagation/num_prop_analysis/num_prop_analysis_energy.py
# SatMAD: Satellite Mission Analysis and Design for Python
#
# Copyright (C) 2020 Egemen Imre
#
# Licensed under GNU GPL v3.0. See LICENSE.rst for more info.
"""
Analysis of numerical propagation errors with various numerical propagation methods.
Case: Energy error and runtimes.
"""
from typing import Dict

from astropy import units as u
from satmad.propagation.numerical_propagators import ODESolverType
from satmad.propagation.tests.num_prop_analysis_engine import (
    analyse_energy_along_trajectory,
    init_rv,
    plot_energy_diff,
    plot_runtimes,
    tle_heo,
)

if __name__ == "__main__":
    # Set up propagation config
    # tle = tle_leo()
    tle = tle_heo()
    rv_init = init_rv(tle, tle.epoch)
    stepsize = 120 * u.s
    solver_types = [solver_type for solver_type in ODESolverType]
    # solver_types.remove(ODESolverType.BDF)
    # solver_types.remove(ODESolverType.RK23)
    # solver_types.clear()
    # solver_types.append(ODESolverType.DOP853)
    rtol = 3e-14
    # atol = [1e-12, 1e-12, 1e-12, 1e-15, 1e-15, 1e-15]
    atol = 1e-12
    init_time_offset = 0 * u.day
    duration = 1.0 * u.day

    cases: Dict[str, Dict] = {}
    type(cases)

    # run all cases
    for solver_type in solver_types:
        name = solver_type.value

        analyse_energy_along_trajectory(
            cases,
            name,
            rv_init,
            stepsize,
            solver_type,
            init_time_offset,
            duration,
            rtol,
            atol,
        )

    session_name = f"[rtol: {rtol}, atol: {atol}]"

    # Plot energy differences
    plot_energy_diff(cases, title=session_name, logscale_plot=True)

    # Plot DOP853 on linear scale to see its behaviour
    plot_energy_diff(
        {"DOP853": cases["DOP853"]}, title=session_name, logscale_plot=False
    )

    # Plot runtimes
    plot_runtimes(
        cases, title=f"{session_name} [duration: {duration.to_value(u.day)} days]"
    )


satmad_applications/analyses/propagation/num_prop_analysis/num_prop_analysis_accuracy.py
# SatMAD: Satellite Mission Analysis and Design for Python
#
# Copyright (C) 2020 Egemen Imre
#
# Licensed under GNU GPL v3.0. See LICENSE.rst for more info.
"""
Analysis of numerical propagation errors with DOP853 numerical propagation method.
Case: Position accuracy
"""
from typing import Dict

from astropy import units as u
from satmad.propagation.numerical_propagators import ODESolverType
from satmad.propagation.tests.num_prop_analysis_engine import (
    analyse_pos_along_trajectory,
    init_rv,
    plot_pos_diff,
    plot_runtimes,
    propagation_engine,
    tle_leo,
)

if __name__ == "__main__":
    # Set up propagation config
    tle = tle_leo()
    rv_init = init_rv(tle, tle.epoch)
    stepsize = 120 * u.s
    solver_type = ODESolverType.DOP853
    rtol = 3e-14
    # atols = [1e-15, 1e-14, 1e-13, 1e-12, 1e-11, 1e-10]
    rtols = [1e-14, 1e-13, 1e-12, 1e-11, 1e-10]
    atol = 1e-14
    init_time_offset = 0 * u.day
    duration = 10.0 * u.day

    # atol effect for fixed low rtol: no impact on pos error
    # atol effect for fixed high rtol: no impact on pos error, lower accuracy
    # rtol effect for fixed low atol: pos error varies with rtol
    # rtol effect for fixed high atol: pos error varies with rtol, same accuracy

    # run propagation and get truth trajectory
    truth_traj = propagation_engine(
        rv_init,
        stepsize,
        ODESolverType.DOP853,
        init_time_offset,
        duration,
        rtol=3e-14,
        atol=1e-15,
    )

    cases: Dict[str, Dict] = {}
    type(cases)

    # run all cases
    for rtol in rtols:
        name = str(rtol)
        # rtol = atol

        analyse_pos_along_trajectory(
            truth_traj,
            cases,
            name,
            rv_init,
            stepsize,
            solver_type,
            init_time_offset,
            duration,
            rtol,
            atol,
        )

    session_name = f"[DOP853, atol = {atol} , varying rtol]"

    # Plot pos diff differences
    plot_pos_diff(cases, title=session_name, logscale_plot=True)

    # Plot runtime for one case to see its behaviour
    plot_pos_diff(
        {"1e-10": cases["1e-10"]}, title=f"{session_name}", logscale_plot=False
    )

    # Plot runtimes
    plot_runtimes(
        cases,
        title=f"{session_name} [duration: {duration.to_value(u.day)} days]",
        logscale_plot=True,
    )


satmad_applications/analyses/propagation/num_prop_analysis/num_prop_analysis_ecc.py
# SatMAD: Satellite Mission Analysis and Design for Python
#
# Copyright (C) 2020 Egemen Imre
#
# Licensed under GNU GPL v3.0. See LICENSE.rst for more info.
"""
Analysis of numerical propagation errors with DOP853 numerical propagation method.
Case: Eccentricity and pos accuracy.
"""
from typing import Dict

from astropy import units as u
from satmad.propagation.numerical_propagators import ODESolverType
from satmad.propagation.tests.num_prop_analysis_engine import (
    analyse_pos_along_trajectory,
    init_rv,
    plot_pos_diff,
    plot_runtimes,
    propagation_engine,
    tle_heo,
)

if __name__ == "__main__":

    # Set up propagation config
    stepsize = 120 * u.s
    solver_type = ODESolverType.DOP853
    rtol = 1e-11
    atol = 1e-11
    init_time_offset = 0 * u.day
    duration = 10.0 * u.day

    ecc_list = [0.001, 0.1, 0.2, 0.4, 0.5, 0.7, 0.75]
    tle_list = []
    for ecc in ecc_list:
        tle = tle_heo()
        tle.eccentricity = ecc
        tle_list.append(tle)

    cases: Dict[str, Dict] = {}
    type(cases)

    # run all cases
    for tle in tle_list:
        name = str(tle.eccentricity)

        rv_init = init_rv(tle, tle.epoch)

        # run propagation and get truth trajectory
        truth_traj = propagation_engine(
            rv_init,
            stepsize,
            ODESolverType.DOP853,
            init_time_offset,
            duration,
            rtol=3e-14,
            atol=1e-15,
        )

        analyse_pos_along_trajectory(
            truth_traj,
            cases,
            name,
            rv_init,
            stepsize,
            solver_type,
            init_time_offset,
            duration,
            rtol,
            atol,
        )

    session_name = f"[DOP853, atol = {atol} , rtol = {rtol}]"

    # Plot pos differences
    plot_pos_diff(cases, title=session_name, logscale_plot=True)

    # Plot runtimes
    plot_runtimes(
        cases,
        title=f"{session_name} [duration: {duration.to_value(u.day)} days]",
        logscale_plot=False,
    )
